Dart/src/com/jetbrains/lang/dart/ide/refactoring/moveFile/DartServerMoveDartFileHandler.java

```diff
@@ -0,0 +1,109 @@
+package com.jetbrains.lang.dart.ide.refactoring.moveFile;
+
+import com.intellij.CommonBundle;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.Messages;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.*;
+import com.intellij.refactoring.move.moveFilesOrDirectories.MoveFileHandler;
+import com.intellij.usageView.UsageInfo;
+import com.intellij.util.IncorrectOperationException;
+import com.jetbrains.lang.dart.DartBundle;
+import com.jetbrains.lang.dart.analyzer.DartAnalysisServerService;
+import com.jetbrains.lang.dart.assists.AssistUtils;
+import com.jetbrains.lang.dart.assists.DartSourceEditException;
+import com.jetbrains.lang.dart.ide.refactoring.status.RefactoringStatus;
+import com.jetbrains.lang.dart.psi.DartFile;
+import com.jetbrains.lang.dart.sdk.DartSdk;
+import org.dartlang.analysis.server.protocol.SourceChange;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+import java.util.Map;
+
+public class DartServerMoveDartFileHandler extends MoveFileHandler {
+
+  @Override
+  public boolean canProcessElement(PsiFile psiFile) {
+    if (!(psiFile instanceof DartFile)) {
+      return false;
+    }
+    final Project project = psiFile.getProject();
+    final DartSdk dartSdk = DartSdk.getDartSdk(project);
+    if (dartSdk == null || !DartAnalysisServerService.isDartSdkVersionForMoveFileRefactoring(dartSdk)) {
+      return false;
+    }
+    DartAnalysisServerService dasService = DartAnalysisServerService.getInstance(project);
+    return dasService.isInIncludedRoots(psiFile.getVirtualFile());
+  }
+
+  @Override
+  public void prepareMovedFile(PsiFile psiFile, PsiDirectory moveDestination, Map<PsiElement, PsiElement> oldToNewMap) {
+    final Project project = psiFile.getProject();
+    final VirtualFile virtualFile = psiFile.getVirtualFile();
+
+    final String oldFilePath = FileUtil.toSystemDependentName(virtualFile.getPath());
+    final String newFilePath = FileUtil.toSystemDependentName(moveDestination.getVirtualFile().getPath()) + "/" + virtualFile.getName();
```

Comment:
Slash not converted :). 
To avoid such bugs we usually try to call `toSystemDependentName()` as late as possible. Usually, it means right before showing a string to a user, in this case, it would mean - right before sending to the server.